cere02/lib.rs

```
@@ -558,13 +573,62 @@ mod ddc {
             let caller = self.env().caller();
             self.only_owner(caller)?;
 
+            self.ddn_statuses.take(&p2p_id);
+
             self.ddc_nodes.take(&p2p_id);
             Self::env().emit_event(DDCNodeRemoved { p2p_id });
 
             Ok(())
         }
     }
 
+    // ---- DDN Statuses ----
+    #[derive(
+        Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode, SpreadLayout, PackedLayout,
+    )]
+    #[cfg_attr(feature = "std", derive(Debug, scale_info::TypeInfo))]
+    pub struct DDNStatus {
+        is_online: bool,
+        total_downtime: u64,
+        reference_timestamp: u64,
+        last_timestamp: u64,
+    }
+
+    impl Ddc {
+        #[ink(message)]
+        pub fn report_ddn_status(&mut self, p2p_id: String, is_online: bool) -> Result<()> {
+            let reporter = self.env().caller();
+            self.only_reporter(&reporter)?;
+
+            let ddn_status = match self.ddn_statuses.get_mut(&p2p_id) {
+                Some(ddn_status) => ddn_status,
+                None => panic!("DDN not found"),
+            };
+
+            let now = Self::env().block_timestamp();
+
+            if !ddn_status.is_online {
+                let last_downtime = now - ddn_status.last_timestamp;
+                ddn_status.total_downtime += last_downtime;
+            }
+
+            ddn_status.last_timestamp = now;
+            ddn_status.is_online = is_online;
+
+            Ok(())
+        }
+
+        #[ink(message)]
+        pub fn get_ddn_status(&self, p2p_id: String) -> DDNStatus {
+            let ddn_status = match self.ddn_statuses.get(&p2p_id) {
+                Some(ddn_status) => ddn_status.clone(),
+                None => panic!("DDN not found"),
```

Let's avoid using panic. Let's just return the error.

---

[done](https://github.com/Cerebellum-Network/cere-enterprise-smart-contracts/pull/31/files#diff-a99b45b57bfdc35f8b28ef154fa4418c459a459aa60dfe0f98d8a350a93fa067R635)